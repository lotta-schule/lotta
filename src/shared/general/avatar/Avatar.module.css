.root {
    border: 1px solid rgb(var(--lotta-accent-grey));
    background-color: rgb(var(--lotta-accent-grey));
    border-radius: 50%;
    padding: 0.5em;
    cursor: pointer;
    display: inline-block;
    background-size: cover;
    box-shadow: 1px 1px 7px #0000002e;
    height: 2em; /* TODO: Property */
    width: 2em; /* TODO: Property */
}
