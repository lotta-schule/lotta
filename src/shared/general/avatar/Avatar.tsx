import * as React from 'react';
import clsx from 'clsx';

import styles from './Avatar.module.css';

export interface AvatarProps {
    /**
     * The avatar image source
     */
    src: string;

    /**
     * The title of the avatar
     */
    title?: string;

    className?: string;

    style?: React.CSSProperties;
}

/**
 * Primary UI shared for userAvatar interaction
 */

export const Avatar = React.memo<AvatarProps>(
    ({ src, title, className, style }) => {
        return (
            <div
                role={'img'}
                title={title}
                className={clsx(styles.root, className)}
                style={{ ...style, backgroundImage: `url(${src})` }}
            />
        );
    }
);
Avatar.displayName = 'Avatar';
