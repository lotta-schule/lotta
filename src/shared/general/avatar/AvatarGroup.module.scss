.root {
    display: inline-flex;
    .avatar {
        margin-right: -8px;
    }
    .overshoot {
        border-radius: 50%;
        background-color: rgb(var(--lotta-accent-grey));
        height: 30px;
        width: 30px;
        z-index: 1;
        font-family: var(--lotta-typography-font-family);
        color: rgb(var(--lotta-text-color));
        box-shadow: 1px 1px 7px #0000002e;
        line-height: 2em;
        text-align: center;
    }
}
