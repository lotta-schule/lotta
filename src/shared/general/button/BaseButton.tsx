import * as React from 'react';
import clsx from 'clsx';

export type BaseButtonProps = {
    /**
     * Wether the button is currently 'active' or 'selected'
     */
    selected?: boolean;

    /**
     * Chose from different styles
     */
    variant?: 'default' | 'fill' | 'borderless' | 'error';

    /**
     * Wether the button should have a width of 100%
     */
    fullWidth?: boolean;

    /**
     * CSS Classname
     */
    className?: string;

    /**
     * CSS Styles
     */
    style?: React.CSSProperties;

    /**
     * Content to show on the button
     */
    children?: any;
} & React.HTMLProps<HTMLButtonElement>;

/**
 * Primary UI shared for userAvatar interaction
 */
export const BaseButton = React.forwardRef<any, BaseButtonProps>(
    (
        {
            children,
            style,
            className,
            variant = 'default',
            fullWidth = false,
            selected = false,
            ...props
        },
        ref
    ) => {
        const ComponentClass = props.href ? 'a' : ('button' as any);
        return React.createElement(
            ComponentClass,
            {
                ref,
                type: props.type ?? 'button',
                role: 'button',
                style,
                ['aria-selected']: selected,
                className: clsx(
                    'lotta-base-button',
                    `variant__${variant}`,
                    { selected, 'full-width': fullWidth },
                    className
                ),
                ...props,
            },
            children
        );
    }
);
BaseButton.displayName = 'BaseButton';
