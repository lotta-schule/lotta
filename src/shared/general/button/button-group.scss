.lotta-button-group {
    display: inline-flex;
    flex-direction: row;
    justify-content: stretch;
    button {
        color: rgba(var(--lotta-primary-color), 1);
        background-color: #fff;
        display: inline-block;
        &:not(:only-child) {
            &:not(:first-of-type):not(:last-of-type):not(.is-first-button-group-button):not(.is-last-button-group-button) {
                border-radius: 0;
                border-right: 0;
            }
            &:first-of-type:not(.is-last-button-group-button),
            &.is-first-button-group-button {
                border-top-right-radius: 0;
                border-bottom-right-radius: 0;
                border-right: 0;
            }
            &:last-of-type:not(.is-first-button-group-button),
            &.is-last-button-group-button {
                border-top-left-radius: 0;
                border-bottom-left-radius: 0;
            }
        }
        &.is-first-button-group-button {
            border-top-right-radius: 0;
            border-bottom-right-radius: 0;
            border-right: 0;
        }
        &.is-last-button-group-button {
            border-top-left-radius: 0;
            border-bottom-left-radius: 0;
        }
        &.full-width {
            flex-grow: 1;
        }
        .lotta-button_icon svg {
            font-size: 1.3rem;
        }
    }
    &.full-width {
        width: 100%;
    }
}
