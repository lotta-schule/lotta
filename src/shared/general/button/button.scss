.lotta-button {
    color: rgba(var(--lotta-primary-color), 1);
    font-weight: 200;
    padding: 0.4em 0.8em;
    align-items: center;
    &:not(.small) {
        min-height: 2.8em;
    }
}

.lotta-button_icon {
    display: flex;
}

.lotta-button_text {
    margin: auto;
    display: contents;
}

aside .lotta-button_text {
    margin: 0;
}

aside .lotta-navigation-button {
    border-radius: 50%;
}

aside nav.lotta-navbar ul {
    height: auto;
}

.lotta-button.only-icon:not(.lotta-button-group .lotta-button) {
    border-radius: 50%;
    border-color: transparent;
    padding: 0.4em;
}

.lotta-button.only-icon
    :not(.lotta-button-group .lotta-button)
    :not(.lotta-button.variant__fill)
    :hover {
    background-color: rgba(var(--lotta-primary-color), 0.2);
}
