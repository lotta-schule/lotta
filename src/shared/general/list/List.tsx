import * as React from 'react';
import { Divider } from '../divider/Divider';
import clsx from 'clsx';

import styles from './List.module.scss';

export type ListProps = React.HTMLProps<HTMLUListElement>;

export const List: React.FC<ListProps> = ({
    children,
    className,
    ...props
}) => {
    return (
        <ul className={clsx(styles.root, className)} {...props}>
            {children}
        </ul>
    );
};
List.displayName = 'List';

export type ListItemProps = (
    | React.HTMLProps<HTMLLIElement>
    | React.HTMLProps<HTMLLinkElement>
) & {
    leftSection?: React.ReactNode;
    rightSection?: React.ReactNode;
    isDivider?: boolean;
};

export const ListItem = React.forwardRef<
    HTMLLinkElement | HTMLLIElement,
    ListItemProps
>(({ children, className, leftSection, rightSection, href, ...props }, ref) => {
    if ('isDivider' in props && props.isDivider === true) {
        const { isDivider, ...rest } =
            props as React.HTMLProps<HTMLLIElement> & { isDivider: boolean };
        return (
            <li
                className={clsx(styles.li, styles.isDivider, className)}
                disabled
                {...rest}
            >
                <Divider />
            </li>
        );
    }
    const elementName = href ? 'a' : 'li';
    return React.createElement(
        elementName,
        { className: clsx(styles.li, className), href, ref, ...props },
        <>
            {leftSection && <div>{leftSection}</div>}
            <div className={styles.mainSection}>{children}</div>
            {rightSection && <div>{rightSection}</div>}
        </>
    );
});
ListItem.displayName = 'ListItem';

export type ListItemSecondaryTextProps = React.HTMLProps<HTMLSpanElement>;

export const ListItemSecondaryText: React.FC<ListItemSecondaryTextProps> = ({
    children,
}) => <span className={styles.listItemSecondaryText}>{children}</span>;
ListItemSecondaryText.displayName = 'ListItemSecondaryText';
