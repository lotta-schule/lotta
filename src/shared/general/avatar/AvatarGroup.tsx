import * as React from 'react';
import { AvatarProps } from './Avatar';
import clsx from 'clsx';

import styles from './AvatarGroup.module.scss';

export interface AvatarGroupProps {
    children: React.ReactElement<AvatarProps>[];

    /**
     * The maximum number of avatars to show
     */
    max?: number;

    className?: string;
}

/**
 * Primary UI shared for userAvatar interaction
 */

export const AvatarGroup: React.FC<AvatarGroupProps> = ({
    max = 3,
    className,
    children,
}) => {
    const avatars = React.Children.toArray(
        children
    ) as React.ReactElement<AvatarProps>[];
    const overshoot = avatars.length - max;
    return (
        <div className={clsx(styles.root, className)}>
            <div role={'group'}>
                {avatars.splice(0, max).map((child, index) => {
                    return React.cloneElement(child, {
                        className: clsx(child.props.className, styles.avatar),
                    });
                })}
            </div>
            {overshoot > 0 && (
                <div className={styles.overshoot}>+{overshoot}</div>
            )}
        </div>
    );
};
