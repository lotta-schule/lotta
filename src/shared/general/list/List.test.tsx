import { render } from 'test/util';
import { Avatar } from '../avatar/Avatar';
import { Button } from '../button/Button';
import { List, ListItem } from './List';

describe('List', () => {
    it('should render', () => {
        const screen = render(
            <List>
                <ListItem>Test</ListItem>
                <ListItem>Test</ListItem>
                <ListItem>Test</ListItem>
                <ListItem>Test</ListItem>
            </List>
        );
        expect(screen.container).toMatchSnapshot();
    });

    it('should render with subsection left', () => {
        const screen = render(
            <List>
                <ListItem leftSection={<Avatar src={''} />}>Test</ListItem>
                <ListItem leftSection={<Avatar src={''} />}>Test</ListItem>
                <ListItem leftSection={<Avatar src={''} />}>Test</ListItem>
                <ListItem leftSection={<Avatar src={''} />}>Test</ListItem>
            </List>
        );
        expect(screen.container).toMatchSnapshot();
    });

    it('should render with subsection right', () => {
        const screen = render(
            <List>
                <ListItem
                    leftSection={<Avatar src={''} />}
                    rightSection={<Button>Test</Button>}
                >
                    Test
                </ListItem>
                <ListItem rightSection={<Button>Test</Button>}>Test</ListItem>
                <ListItem rightSection={<Button>Test</Button>}>Test</ListItem>
                <ListItem rightSection={<Button>Test</Button>}>Test</ListItem>
            </List>
        );
        expect(screen.container).toMatchSnapshot();
    });
});
