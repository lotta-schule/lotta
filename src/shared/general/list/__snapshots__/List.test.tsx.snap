// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render 1`] = `
<div>
  <div
    data-overlay-container="true"
  >
    <ul
      class="root"
    >
      <li
        class="li"
      >
        <div
          class="mainSection"
        >
          Test
        </div>
      </li>
      <li
        class="li"
      >
        <div
          class="mainSection"
        >
          Test
        </div>
      </li>
      <li
        class="li"
      >
        <div
          class="mainSection"
        >
          Test
        </div>
      </li>
      <li
        class="li"
      >
        <div
          class="mainSection"
        >
          Test
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`List should render with subsection left 1`] = `
<div>
  <div
    data-overlay-container="true"
  >
    <ul
      class="root"
    >
      <li
        class="li"
      >
        <div>
          <div
            class="root"
            role="img"
            style="background-image: url();"
          />
        </div>
        <div
          class="mainSection"
        >
          Test
        </div>
      </li>
      <li
        class="li"
      >
        <div>
          <div
            class="root"
            role="img"
            style="background-image: url();"
          />
        </div>
        <div
          class="mainSection"
        >
          Test
        </div>
      </li>
      <li
        class="li"
      >
        <div>
          <div
            class="root"
            role="img"
            style="background-image: url();"
          />
        </div>
        <div
          class="mainSection"
        >
          Test
        </div>
      </li>
      <li
        class="li"
      >
        <div>
          <div
            class="root"
            role="img"
            style="background-image: url();"
          />
        </div>
        <div
          class="mainSection"
        >
          Test
        </div>
      </li>
    </ul>
  </div>
</div>
`;

exports[`List should render with subsection right 1`] = `
<div>
  <div
    data-overlay-container="true"
  >
    <ul
      class="root"
    >
      <li
        class="li"
      >
        <div>
          <div
            class="root"
            role="img"
            style="background-image: url();"
          />
        </div>
        <div
          class="mainSection"
        >
          Test
        </div>
        <div>
          <button
            aria-selected="false"
            class="lotta-base-button variant__default lotta-button"
            role="button"
            type="button"
          >
            <span
              class="lotta-button_text"
            >
              Test
            </span>
          </button>
        </div>
      </li>
      <li
        class="li"
      >
        <div
          class="mainSection"
        >
          Test
        </div>
        <div>
          <button
            aria-selected="false"
            class="lotta-base-button variant__default lotta-button"
            role="button"
            type="button"
          >
            <span
              class="lotta-button_text"
            >
              Test
            </span>
          </button>
        </div>
      </li>
      <li
        class="li"
      >
        <div
          class="mainSection"
        >
          Test
        </div>
        <div>
          <button
            aria-selected="false"
            class="lotta-base-button variant__default lotta-button"
            role="button"
            type="button"
          >
            <span
              class="lotta-button_text"
            >
              Test
            </span>
          </button>
        </div>
      </li>
      <li
        class="li"
      >
        <div
          class="mainSection"
        >
          Test
        </div>
        <div>
          <button
            aria-selected="false"
            class="lotta-base-button variant__default lotta-button"
            role="button"
            type="button"
          >
            <span
              class="lotta-button_text"
            >
              Test
            </span>
          </button>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
